src/fluentd/src/fluent-plugin-pgjson/lib/fluent/plugin/out_pgjson.rb

```diff
@@ -140,8 +140,65 @@ def format(tag, time, record)
+      historyType = "retry"
+      snapshot = record_value(record["objectSnapshot"])
+      # use frameworkName + attemptIndex + historyType to generate a uid
+      uid = Digest::MD5.hexdigest "#{frameworkName}+#{attemptIndex}+#{historyType}"
+      # select from framework_history, ensure there is no corresponding history object
+      selectResult = thread[:conn].exec_params('SELECT uid from framework_history where uid=$1', [uid])
+      if selectResult.cmd_tuples == 0
+          # if there is no existing records, try to insert a new one.
+          thread[:conn].exec_params("INSERT INTO framework_history (\"#{@insertedAt_col}\", \"#{@updatedAt_col}\", \"#{@uid_col}\", \"#{@frameworkName_col}\", \"#{@attemptIndex_col}\", \"#{@historyType_col}\", \"#{@snapshot_col}\") " +
+            "VALUES ($1, $2, $3, $4, $5, $6, $7)", [time, time, uid, frameworkName, attemptIndex, historyType, snapshot]
+          )
+      else
+        # if there is an existing record, ignore it.
+        log.warn "[pgjson] chunk #{hex_id}: ignored framework snapshot object, uid=#{uid}"
+      end
+    end
+
+    def insert_pod(hex_id, time, record)
+      # This function try to insert the pod snapshot into pods table.
+      # In some cases, the framework controller may have duplicate logs about one pod,
+      # or there has been already successful inserted record before.
+      # To handle it, `insert_pod` executes a "SELECT" first.
+      # If the uid exists in the table, it will ignore it safely.
+      # Any error should be raised.
+      thread = Thread.current
+      uid = record["objectSnapshot"]["metadata"]["uid"]
+      frameworkName = record["objectSnapshot"]["metadata"]["name"][0..31]
+      attemptIndex = record["objectSnapshot"]["metadata"]["annotations"]["FC_FRAMEWORK_ATTEMPT_ID"]
+      taskroleName = record["objectSnapshot"]["metadata"]["annotations"]["FC_TASKROLE_NAME"]
+      taskroleIndex = record["objectSnapshot"]["metadata"]["annotations"]["FC_TASK_INDEX"]
+      taskAttemptIndex = record["objectSnapshot"]["metadata"]["annotations"]["FC_TASK_ATTEMPT_ID"]
+      snapshot = record_value(record["objectSnapshot"])
+      selectResult = thread[:conn].exec_params('SELECT uid from pods where uid=$1', [uid])
+      if selectResult.cmd_tuples == 0
+          # if there is no existing records, try to insert a new one.
+          thread[:conn].exec("INSERT INTO pods (\"#{@insertedAt_col}\", \"#{@updatedAt_col}\", \"#{@uid_col}\", \"#{@frameworkName_col}\", \"#{@attemptIndex_col}\", \"#{@taskroleName_col}\", \"#{@taskroleIndex_col}\", \"#{@taskAttemptIndex_col}\", \"#{@snapshot_col}\") " + 
+            "VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)", [time, time, uid, frameworkName, attemptIndex, taskroleName, taskroleIndex, taskAttemptIndex, snapshot]
+          )
+      else
+        # if there is an existing record, ignore it.
+        log.warn "[pgjson] chunk #{hex_id}: ignored pod snapshot object, uid=#{uid}"
```

Comment:
log.warn "[pgjson] chunk #{hex_id}: ignored pod snapshot object as it already exist, uid=#{uid}"

Reply:
Done